File: trailnest-api/src/main/java/com/trailnest/api/domain/Campground.java
package com.trailnest.api.domain;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.BatchSize;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "campgrounds")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Campground {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, unique = true, length = 255)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(precision = 10, scale = 2)
    private BigDecimal price;

    @Column(nullable = false, length = 500)
    private String location;

    // coordinates (GEOGRAPHY) is intentionally not mapped here.
    // It is set via a native UPDATE query in CampgroundRepository
    // and read via native queries for geospatial searches.

    @Column(name = "elevation_m")
    private Integer elevationM;

    @Builder.Default
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb", nullable = false)
    private List<String> amenities = new ArrayList<>();

    @Column(name = "campground_type", length = 50)
    private String campgroundType;

    @Column(name = "max_capacity")
    private Integer maxCapacity;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @Builder.Default
    @Column(name = "avg_rating", precision = 3, scale = 2, nullable = false)
    private BigDecimal avgRating = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "review_count", nullable = false)
    private Integer reviewCount = 0;

    @Builder.Default
    @Column(name = "is_published", nullable = false)
    private boolean published = true;

    @Builder.Default
    @Column(name = "is_featured", nullable = false)
    private boolean featured = false;

    // Batched so loading images for N campgrounds issues ceil(N/20) queries, not N
    @BatchSize(size = 20)
    @OneToMany(mappedBy = "campground", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("displayOrder ASC")
    @Builder.Default
    private List<CampgroundImage> images = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = updatedAt = OffsetDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}

File: trailnest-web/src/types/campground.ts
export interface CampgroundImage {
	id: string;
	url: string;
	primary: boolean;
	displayOrder: number;
}

export interface CampgroundAuthor {
	id: string;
	email: string;
	username: string;
	displayName: string | null;
	avatarUrl: string | null;
	role: string;
	verified: boolean;
}

/** Full detail — returned by GET /campgrounds/:slug */
export interface Campground {
	id: string;
	title: string;
	slug: string;
	description: string | null;
	price: number | null;
	location: string;
	latitude: number | null;
	longitude: number | null;
	elevationM: number | null;
	amenities: string[];
	campgroundType: string | null;
	maxCapacity: number | null;
	avgRating: number;
	reviewCount: number;
	published: boolean;
	featured: boolean;
	images: CampgroundImage[];
	author: CampgroundAuthor;
	createdAt: string;
	updatedAt: string;
}

/** Lightweight card — returned by GET /campgrounds (list) */
export interface CampgroundSummary {
	id: string;
	title: string;
	slug: string;
	description: string | null;
	price: number | null;
	location: string;
	avgRating: number;
	reviewCount: number;
	campgroundType: string | null;
	primaryImageUrl: string | null;
	createdAt: string;
}

export interface CreateCampgroundRequest {
	title: string;
	description?: string;
	price?: number;
	location: string;
	latitude?: number;
	longitude?: number;
	amenities?: string[];
	campgroundType?: string;
	maxCapacity?: number;
}

export interface UpdateCampgroundRequest extends Partial<CreateCampgroundRequest> {
	published?: boolean;
}
